zoo/board_games/mcts_bot.py

```diff
@@ -177,52 +237,73 @@ def best_action(self, simulations_number=None, total_simulation_seconds=None):
             Returns the best children node, and can get action from Node.best_action_index.
         """
 
+        # The search cost is determined by either the maximum number of simulations or the longest simulation time.
         if simulations_number is None:
             assert (total_simulation_seconds is not None)
             end_time = time.time() + total_simulation_seconds
             while True:
+                # get the leaf node.
                 v = self._tree_policy()
+                # rollout from the leaf node.
                 reward = v.rollout()
+                # backpropagate from the leaf node to the root node.
                 v.backpropagate(reward)
                 if time.time() > end_time:
                     break
         else:
             for i in range(0, simulations_number):
                 # print('****simlulation-{}****'.format(i))
+                # get the leaf node.
                 v = self._tree_policy()
+                # rollout from the leaf node.
                 reward = v.rollout()
                 # print('reward={}\n'.format(reward))
+                # backpropagate from the leaf node to the root node.
                 v.backpropagate(reward)
         # to select best child go for exploitation only
+        # 这里和算法理论里选择最高访问次数似乎不太一样？？？？？？？？？？？？？？？？？？
         return self.root.best_child(c_param=0.)
 
+    #
     def _tree_policy(self):
         """
         Overview:
-            The policy used to select action to rollout a episode.
+            This function implement the tree search from the root node to the leaf node, which is visited for the first time or is the terminal node.
         """
         current_node = self.root
         while not current_node.is_terminal_node():
             if not current_node.is_fully_expanded():
+                # choose a child node which has not been visited before
                 return current_node.expand()
             else:
                 current_node = current_node.best_child()
         return current_node
 
 
 class MCTSBot:
-
+    """
+    Overview:
+        A robot which can use MCTS to make decision, choose an action to take.
+    """
+ 
     def __init__(self, ENV, cfg, bot_name, num_simulation=10000):
         self.name = bot_name
         self.num_simulation = num_simulation
         self.ENV = ENV
         self.cfg = cfg
 
     def get_actions(self, state, player_index):
```

Comment:
"""
         Get the best action for a given game state.

         A new environment is created for simulations. The environment is reset to the given state.
         Then, MCTS is performed with the specified number of simulations to find the best action.

         Arguments:
             - state: The current game state.
             - player_index (int): The index of the current player.

         Returns:
             The best action determined by MCTS.
         """